File: src/lib/tauri/proxy.ts
/**
 * tauri/proxy — Local Proxy domain facade（ADR-0020 / plan3 UI-001）。
 */

import { cmd } from './core';

export type PortMode = 'dynamic' | 'fixed';

export interface ProxySettings {
  id: string;
  enabledIntent: boolean;
  bindHost: string;
  portMode: PortMode;
  configuredPort: number;
  effectivePort: number;
  accessSecretRef: string;
  graceTimeoutMs: number;
  maxConcurrency: number;
  maxRequestBodyBytes: number;
  updatedAt: string;
}

export type ProxyRuntimeStatus =
  | 'stopped'
  | 'starting'
  | 'running'
  | 'restarting'
  | 'stopping'
  | 'failed';

export interface ProxyEndpointInfo {
  protocol: string;
  path: string;
  url: string;
}

export interface ProxyStatusDTO {
  running: boolean;
  status: ProxyRuntimeStatus;
  host: string;
  port: number;
  effectivePort: number;
  startedAt?: string | null;
  uptimeSeconds: number;
  activeRequests: number;
  routeCount: number;
  engine: string;
  lastError?: string | null;
  protocolEndpoints: ProxyEndpointInfo[];
}

export type PoolPolicy = 'priority_round_robin' | 'round_robin' | 'least_inflight';

export type CredentialSelector =
  | { credential: { id: string } }
  | { pool: { policy: PoolPolicy } };

export interface RouteTarget {
  id: string;
  routeId: string;
  position: number;
  connectionId: string;
  modelId: string;
  credentialSelector: CredentialSelector;
  priority: number;
  enabled: boolean;
}

export interface Route {
  id: string;
  localModel: string;
  enabled: boolean;
  strategy: string;
  targets: RouteTarget[];
  createdAt: string;
  updatedAt: string;
}

export interface ProxyUsageRecord {
  id: string;
  routeId?: string | null;
  connectionId?: string | null;
  credentialId?: string | null;
  inboundProtocol: string;
  upstreamProtocol: string;
  localModel: string;
  upstreamModel: string;
  promptTokens: number;
  completionTokens: number;
  totalTokens: number;
  reasoningTokens?: number | null;
  cachedTokens?: number | null;
  latencyMs: number;
  status: string;
  errorCode?: string | null;
  createdAt: string;
}

export interface ProxyChatInput {
  protocol: 'anthropic_messages' | 'openai_chat_completions' | 'openai_responses';
  baseUrl: string;
  secretRef: string;
  model: string;
  requestJson: string;
}

export interface ProxyChatResult {
  ok: boolean;
  usageInputTokens: number;
  usageOutputTokens: number;
  stopReason: string;
  errorCategory?: string | null;
  errorCode?: string | null;
  errorMessage?: string | null;
}

export const proxyApi = {
  status: () => cmd<ProxyStatusDTO>('proxy_status'),
  start: () => cmd<boolean>('proxy_start'),
  stop: () => cmd<boolean>('proxy_stop'),
  restart: () => cmd<boolean>('proxy_restart'),

  getSettings: () => cmd<ProxySettings>('proxy_get_settings'),
  updateSettings: (input: {
    enabledIntent?: boolean;
    bindHost?: string;
    portMode?: string;
    configuredPort?: number;
    graceTimeoutMs?: number;
    maxConcurrency?: number;
    maxRequestBodyBytes?: number;
  }) => cmd<ProxySettings>('proxy_update_settings', { input }),

  listRoutes: () => cmd<Route[]>('proxy_list_routes'),
  getRoute: (id: string) => cmd<Route | null>('proxy_get_route', { id }),
  saveRoute: (route: Route) => cmd<Route>('proxy_save_route', { route }),
  deleteRoute: (id: string) => cmd<boolean>('proxy_delete_route', { id }),

  listUsageRecords: (limit?: number, offset?: number) =>
    cmd<ProxyUsageRecord[]>('proxy_list_usage_records', { limit, offset }),

  chat: (input: ProxyChatInput) => cmd<ProxyChatResult>('proxy_chat', { input }),
};

// 兼容别名
export const proxy = proxyApi;
export type ProxyStatusResult = ProxyStatusDTO;

File: src-tauri/src/commands/proxy.rs
//! Local Proxy Host 命令（ADR-0020 §4 / PRX-001..004 / UI-004..005）。

use serde::{Deserialize, Serialize};
use tauri::State;

use crate::proxy::engine::{EngineCall, ProxyEngine};
use crate::proxy::model::{PortMode, ProxySettings, ProxyStatusDTO, ProxyUsageRecord, Route};
use crate::proxy::native::NativeProxyEngine;
use crate::proxy::store;
use crate::secrets::store::SecretRef;
use crate::AppState;
use crate::{Error, Result};

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdateProxySettingsInput {
    pub enabled_intent: Option<bool>,
    pub bind_host: Option<String>,
    pub port_mode: Option<String>,
    pub configured_port: Option<u16>,
    pub grace_timeout_ms: Option<u64>,
    pub max_concurrency: Option<u32>,
    pub max_request_body_bytes: Option<usize>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProxyChatInput {
    pub protocol: String,
    pub base_url: String,
    pub secret_ref: String,
    pub model: String,
    pub request_json: String,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProxyChatResult {
    pub ok: bool,
    pub usage_input_tokens: u64,
    pub usage_output_tokens: u64,
    pub stop_reason: String,
    pub error_category: Option<String>,
    pub error_code: Option<String>,
    pub error_message: Option<String>,
}

#[tauri::command]
pub fn proxy_get_settings(state: State<'_, AppState>) -> Result<ProxySettings> {
    let conn = state.db.get().map_err(|e| Error::Internal(e.to_string()))?;
    store::get_proxy_settings(&conn)
}

#[tauri::command]
pub fn proxy_update_settings(
    state: State<'_, AppState>,
    input: UpdateProxySettingsInput,
) -> Result<ProxySettings> {
    let conn = state.db.get().map_err(|e| Error::Internal(e.to_string()))?;
    let mut current = store::get_proxy_settings(&conn)?;

    if let Some(enabled) = input.enabled_intent {
        current.enabled_intent = enabled;
    }
    if let Some(host) = input.bind_host {
        current.bind_host = host;
    }
    if let Some(pm) = input.port_mode {
        current.port_mode = PortMode::from_str(&pm);
    }
    if let Some(port) = input.configured_port {
        current.configured_port = port;
    }
    if let Some(grace) = input.grace_timeout_ms {
        current.grace_timeout_ms = grace;
    }
    if let Some(mc) = input.max_concurrency {
        current.max_concurrency = mc;
    }
    if let Some(mb) = input.max_request_body_bytes {
        current.max_request_body_bytes = mb;
    }

    store::save_proxy_settings(&conn, &current)?;
    Ok(current)
}

#[tauri::command]
pub fn proxy_status(state: State<'_, AppState>) -> Result<ProxyStatusDTO> {
    let conn = state.db.get().map_err(|e| Error::Internal(e.to_string()))?;
    state.proxy_runtime.get_status_dto(&conn)
}

#[tauri::command]
pub fn proxy_start(state: State<'_, AppState>) -> Result<bool> {
    state.proxy_runtime.start(state.db.clone()).map(|_| true)
}

#[tauri::command]
pub fn proxy_stop(state: State<'_, AppState>) -> Result<bool> {
    let conn = state.db.get().map_err(|e| Error::Internal(e.to_string()))?;
    state.proxy_runtime.stop(&conn).map(|_| true)
}

#[tauri::command]
pub fn proxy_restart(state: State<'_, AppState>) -> Result<bool> {
    state.proxy_runtime.restart(state.db.clone()).map(|_| true)
}

// ── Routes ──

#[tauri::command]
pub fn proxy_list_routes(state: State<'_, AppState>) -> Result<Vec<Route>> {
    let conn = state.db.get().map_err(|e| Error::Internal(e.to_string()))?;
    store::list_routes(&conn)
}

#[tauri::command]
pub fn proxy_get_route(state: State<'_, AppState>, id: String) -> Result<Option<Route>> {
    let conn = state.db.get().map_err(|e| Error::Internal(e.to_string()))?;
    store::get_route(&conn, &id)
}

#[tauri::command]
pub fn proxy_save_route(state: State<'_, AppState>, route: Route) -> Result<Route> {
    let conn = state.db.get().map_err(|e| Error::Internal(e.to_string()))?;
    store::save_route(&conn, &route)
}

#[tauri::command]
pub fn proxy_delete_route(state: State<'_, AppState>, id: String) -> Result<bool> {
    let conn = state.db.get().map_err(|e| Error::Internal(e.to_string()))?;
    store::delete_route(&conn, &id)
}

// ── Usage ──

#[tauri::command]
pub fn proxy_list_usage_records(
    state: State<'_, AppState>,
    limit: Option<usize>,
    offset: Option<usize>,
) -> Result<Vec<ProxyUsageRecord>> {
    let conn = state.db.get().map_err(|e| Error::Internal(e.to_string()))?;
    store::list_usage_records(&conn, limit.unwrap_or(50), offset.unwrap_or(0))
}

// ── Internal proxy chat ──

#[tauri::command]
pub async fn proxy_chat(
    _state: State<'_, AppState>,
    input: ProxyChatInput,
) -> Result<ProxyChatResult> {
    let store = crate::secrets::keychain::KeychainSecretStore::default();
    let engine = NativeProxyEngine::new(std::sync::Arc::new(store));

    let reference = SecretRef::new(input.secret_ref);

    match engine
        .call(EngineCall {
            protocol: input.protocol,
            base_url: input.base_url,
            secret_ref: reference,
            model: input.model,
            request_json: input.request_json,
        })
        .await
    {
        Ok(crate::proxy::engine::EngineOutcome::Completed {
            usage_input_tokens,
            usage_output_tokens,
            stop_reason,
        }) => Ok(ProxyChatResult {
            ok: true,
            usage_input_tokens,
            usage_output_tokens,
            stop_reason,
            error_category: None,
            error_code: None,
            error_message: None,
        }),
        Ok(crate::proxy::engine::EngineOutcome::Failed {
            category,
            code,
            retryable: _,
        }) => Ok(ProxyChatResult {
            ok: false,
            usage_input_tokens: 0,
            usage_output_tokens: 0,
            stop_reason: String::new(),
            error_category: Some(category),
            error_code: Some(code),
            error_message: None,
        }),
        Ok(crate::proxy::engine::EngineOutcome::Unavailable { reason }) => Ok(ProxyChatResult {
            ok: false,
            usage_input_tokens: 0,
            usage_output_tokens: 0,
            stop_reason: String::new(),
            error_category: Some("unavailable".into()),
            error_code: Some("engine_unavailable".into()),
            error_message: Some(reason),
        }),
        Err(error) => Ok(ProxyChatResult {
            ok: false,
            usage_input_tokens: 0,
            usage_output_tokens: 0,
            stop_reason: String::new(),
            error_category: Some(error.category),
            error_code: Some(error.code),
            error_message: Some(error.message),
        }),
    }
}
